File: src/content/3.2/code/scala/snippet05.scala
abstract class Adjunction[F[_], U[_]](
  implicit val F: Functor[F],
  val U: Representable[U]
){
  // changed the order of parameters
  // to help Scala compiler infer types
  def leftAdjunct[A, B](a: A)(f: F[A] => B): U[B]

  def rightAdjunct[A, B](fa: F[A])(f: A => U[B]): B
}

File: src/content/3.2/code/ocaml/snippet05.ml
module type Adjunction_HomSet = functor
  (F : Functor)
  (U : Representable)
  -> sig
  val left_adjunct : ('a F.t -> 'b) -> 'a -> 'b U.t
  val right_adjunct : ('a -> 'b U.t) -> 'a F.t -> 'b
end
